sklearn/utils/stats.py

```diff
@@ -79,112 +79,133 @@ def _weighted_percentile(
 
     Returns
     -------
-    percentile : scalar or 0D array if `array` 1D (or 0D), array if `array` 2D
-        Weighted percentile at the requested probability level.
+    percentile : scalar, 1D array, or 2D array
+        Weighted percentile at the requested probability level(s).
+        If `array` is 1D and `percentile_rank` is scalar, returns a scalar.
+        If `array` is 2D and `percentile_rank` is scalar, returns a 1D array
+            of shape `(array.shape[1],)`
+        If `array` is 1D and `percentile_rank` is 1D, returns a 1D array
+            of shape `(percentile_rank.shape[0],)`
+        If `array` is 2D and `percentile_rank` is 1D, returns a 2D array
+            of shape `(array.shape[1], percentile_rank.shape[0])`
     """
     xp, _, device = get_namespace_and_device(array)
     # `sample_weight` should follow `array` for dtypes
+    # XXX: ^ why? Also: why floating dtype? (is this really floating, I don't think so)
```

Comment:
Since `average=True` can return the average between observed values, the result can be a floating point value even if the input data dtype is integer-based. E.g. the median of `[1, 2]` is 1.5 with `average=True` and unit weights.

Or maybe you remark is about the "floating dtype" expression itself?  "floating dtype" was meant as a short-hand for "floating point dtype". If that helps, `_find_matching_floating_dtype` could be renamed `_find_matching_floating_point_dtype` to avoid the confusion.